File: js/src/proxy.ts
export interface ProxySettings {
  proxy: {
    server: string;
    username?: string;
    password?: string;
  };
}

export function parseProxySettings(proxyUrl: string): ProxySettings {
  if (!proxyUrl.includes("://")) {
    throw new Error(
      `Invalid proxy URL: ${proxyUrl}. Expected format: http(s)://host:port`
    );
  }

  const parsed = new URL(proxyUrl);
  const scheme = parsed.protocol.slice(0, -1).toLowerCase();
  if (scheme !== "http" && scheme !== "https") {
    throw new Error(
      `Unsupported proxy scheme: ${scheme}. Only http:// and https:// proxies are supported.`
    );
  }
  if (!parsed.hostname) {
    throw new Error(
      `Invalid proxy URL: ${proxyUrl}. Expected format: ${scheme}://host:port`
    );
  }

  const hostPort = parsed.port ? `${parsed.hostname}:${parsed.port}` : parsed.hostname;
  const proxy: ProxySettings["proxy"] = { server: `${scheme}://${hostPort}` };
  const username = parsed.username ? decodeURIComponent(parsed.username) : "";
  const password = parsed.password ? decodeURIComponent(parsed.password) : "";

  if (parsed.username) {
    proxy.username = username;
    proxy.password = password;
  }

  return { proxy };
}

File: src/camoufox_cli/proxy.py
"""Proxy parsing helpers shared by browser launch code and tests."""

from __future__ import annotations

from urllib.parse import unquote, urlparse


def parse_proxy_settings(proxy_url: str) -> dict[str, str]:
    parsed = urlparse(proxy_url)
    if parsed.scheme not in ("http", "https"):
        raise ValueError(
            f"Unsupported proxy scheme: {parsed.scheme}. Only http:// and https:// proxies are supported."
        )
    if not parsed.hostname:
        raise ValueError(
            f"Invalid proxy URL: {proxy_url}. Expected format: {parsed.scheme}://host:port"
        )

    host_port = parsed.hostname
    if parsed.port:
        host_port += f":{parsed.port}"

    proxy = {"server": f"{parsed.scheme}://{host_port}"}
    username = unquote(parsed.username) if parsed.username is not None else None
    password = unquote(parsed.password) if parsed.password is not None else ""
    if username is not None:
        proxy["username"] = username
        proxy["password"] = password

    return proxy
